File: clients/desktop/examples/go-examples/websocket/filesend.go
package main

import (
	"encoding/json"
	"fmt"
	"github.com/btcsuite/btcutil/base58"
	"github.com/gorilla/websocket"
	"io/ioutil"
)

func getSelfAddress(conn *websocket.Conn) string {
	selfAddressRequest, err := json.Marshal(map[string]string{"type": "selfAddress"})
	if err != nil {
		panic(err)
	}

	if err = conn.WriteMessage(websocket.TextMessage, []byte(selfAddressRequest)); err != nil {
		panic(err)
	}

	responseJSON := make(map[string]interface{})
	err = conn.ReadJSON(&responseJSON)
	if err != nil {
		panic(err)
	}

	return responseJSON["address"].(string)
}

func main() {
	uri := "ws://localhost:1977"

	conn, _, err := websocket.DefaultDialer.Dial(uri, nil)
	if err != nil {
		panic(err)
	}
	defer conn.Close()

	selfAddress := getSelfAddress(conn)
	fmt.Printf("our address is: %v\n", selfAddress)
	decodedAddress := base58.Decode(selfAddress)

	read_data, err := ioutil.ReadFile("dummy_file")
	if err != nil {
		panic(err)
	}

	payload := append(decodedAddress[:], read_data[:]...)
	fmt.Printf("sending content of 'dummy file' over the mix network...\n")
	if err = conn.WriteMessage(websocket.BinaryMessage, payload); err != nil {
		panic(err)
	}
	sendConfirmationJSON := make(map[string]interface{})
	err = conn.ReadJSON(&sendConfirmationJSON)
	if err != nil {
		panic(err)
	}
	if sendConfirmationJSON["type"].(string) != "send" {
		panic("invalid send confirmation")
	}

	fmt.Printf("waiting to receive a message from the mix network...\n")
	_, receivedMessage, err := conn.ReadMessage()
	if err != nil {
		panic(err)
	}

	fmt.Printf("writing the file back to the disk!\n")
	ioutil.WriteFile("received_file", receivedMessage, 0644)
}

File: clients/desktop/examples/python-examples/websocket/filesend.py

import asyncio
import base58
import json
import websockets

self_address_request = json.dumps({
    "type": "selfAddress"
})


async def send_file():
    uri = "ws://localhost:1977"
    async with websockets.connect(uri) as websocket:
        await websocket.send(self_address_request)
        self_address = json.loads(await websocket.recv())
        print("our address is: {}".format(self_address["address"]))

        bin_payload = bytearray(base58.b58decode(self_address["address"]))
        with open("dummy_file", "rb") as input_file:
            read_data = input_file.read()
            bin_payload += read_data

        print("sending content of 'dummy_file' over the mix network...")
        await websocket.send(bin_payload)
        msg_send_confirmation = json.loads(await websocket.recv())
        assert msg_send_confirmation["type"], "send"

        print("waiting to receive the 'dummy_file' from the mix network...")
        received_data = await websocket.recv()
        with open("received_file", "wb") as output_file:
            print("writing the file back to the disk!")
            output_file.write(received_data)

asyncio.get_event_loop().run_until_complete(send_file())
